File: src/simulation/space/resource/dependency/LabelerDependency.java
package simulation.space.resource.dependency;

import simulation.space.resource.tag.labeler.ResourceLabeler;
import simulation.space.tile.Tile;
import simulation.space.resource.Resource;

public abstract class LabelerDependency extends CoefficientDependency {
    private ResourceLabeler goodResource;
    private boolean isNecessary;
    double amount;

    public LabelerDependency(
            double deprivationCoefficient,
            boolean isNecessary,
            double amount,
            ResourceLabeler goodResource
    ) {
        super(deprivationCoefficient);
        this.isNecessary = isNecessary;
        this.amount = amount;
        this.goodResource = goodResource;
    }

    @Override
    public boolean isNecessary() {
        return isNecessary;
    }

    @Override
    public boolean isResourceNeeded() {
        return true;
    }

    @Override
    public boolean hasNeeded(Tile tile) {
        return tile.getResourcePack().any(this::isResourceGood);
    }

    boolean isResourceGood(Resource resource) {
        return isResourceDependency(resource);
    }

    public boolean isResourceDependency(Resource resource) {
        return goodResource.isSuitable(resource.getGenome()) && resource.isNotEmpty();
    }

    public int oneResourceWorth(Resource resource) {
        return goodResource.actualMatches(resource.copy(1, resource.getOwnershipMarker())).stream()
                .map(Resource::getAmount)
                .reduce(0, Integer::sum);
    }

    public int partByResource(Resource resource, double amount) {
        return (int) Math.ceil(amount / goodResource.actualMatches(resource.copy(1, resource.getOwnershipMarker())).stream()
                .map(Resource::getAmount)
                .reduce(0, Integer::sum));
    }
}

File: src/simulation/space/resource/dependency/LabelerDependency.kt
package simulation.space.resource.dependency

import simulation.space.resource.Resource
import simulation.space.resource.tag.labeler.ResourceLabeler
import simulation.space.tile.Tile
import kotlin.math.ceil


abstract class LabelerDependency(
        deprivationCoefficient: Double,
        override val isNecessary: Boolean,
        var amount: Double,
        private val goodResource: ResourceLabeler
) : CoefficientDependency(deprivationCoefficient) {
    override val isResourceNeeded = true

    override fun hasNeeded(tile: Tile) = tile.resourcePack.any { isResourceGood(it) }

    fun isResourceGood(resource: Resource) = isResourceDependency(resource)

    open fun isResourceDependency(resource: Resource) =
            goodResource.isSuitable(resource.genome) && resource.isNotEmpty

    fun oneResourceWorth(resource: Resource) =
            goodResource.actualMatches(resource.copy(1, resource.ownershipMarker))
                    .map(Resource::amount)
                    .foldRight(0, Int::plus)

    fun partByResource(resource: Resource, amount: Double) = ceil(
            amount / goodResource.actualMatches(resource.copy(1, resource.ownershipMarker))
            .map(Resource::amount)
            .foldRight(0, Int::plus)
    ).toInt()
}
